File: java/src/main/java/br/com/emendes/problems/MinimumWindowSubstring.java
package br.com.emendes.problems;

/**
 * Dados duas Strings s e t de comprimentos m e n respectivamente, retorne o minimum window substring de s
 * de modo que cada caractere em t (incluindo duplicatas) seja incluído na janela.
 * Se não existir tal substring, retorne uma String vazia ("").<br><br>
 * <p>
 * O casos de testes são gerados de modo que a resposta seja única.<br><br>
 * <p>
 * Restrições:<br>
 * m == s.length<br>
 * n == t.length<br>
 * 1 <= m, n <= 10⁵<br>
 * s e t consistem de letras maiúsculas e minúculas inglesas.
 */
public class MinimumWindowSubstring {

  public String minWindow(String s, String t) {
    int startAns = -1;
    int endAns = -1;

    int m = s.length();
    int n = t.length();

    for (int i = 0; m - i >= n; i++) {
      boolean[] check = new boolean[n];
      int counter = 0;
      int j;
      for (j = i; j < m && counter < n; j++) {
        char c = s.charAt(j);
        int position = t.indexOf(c);
        while (position != -1 && check[position]) {
          position = t.indexOf(c, position + 1);
        }
        if (position != -1) {
          check[position] = true;
          counter++;
        }
      }

      if (counter == n && (startAns == -1 || endAns - startAns > j - i)) {
        startAns = i;
        endAns = j;
      }
    }

    return startAns == -1 ? "" : s.substring(startAns, endAns);
  }

}

File: javascript/problems/MinimumWindowSubstring.js
/**
 * Dado duas strings s e t de comprimentos m e n respectivamente, retorne a substring de janela mínima
 * de s de modo que todos os elementos em t (incluindo duplicatas) estejam inclusos na janela. Se não existe
 * tal substring, retorne uma string vazia.
 *
 * Os casos de testes serão gerados de modo que a resposta seja única.
 *
 * Restrições:
 *
 * m == s.length
 *
 * n = t.length
 *
 * 1 <= m, n <= 10⁵
 *
 * s e t consistem de letras maiúsculas.
 */
const minWindow = function (s, t) {
  const counter = new Array(128).fill(0);
  for (let index = 0; index < t.length; index++) counter[t.charCodeAt(index)]++;
  let required = t.length;
  let start = -1;
  let end = s.length;
  for (let l = 0, r = 0; r < s.length; r++) {
    if (--counter[s.charCodeAt(r)] >= 0) required--;
    while (required === 0) {
      if (r - l + 1 < end - start) {
        start = l;
        end = r + 1;
      }
      if (++counter[s.charCodeAt(l)] > 0) required++;
      l++;
    }
  }

  return start === -1 ? "" : s.substring(start, end);
};

const firstSolution = function (s, t) {
  let answer = [0, 1000000];
  const tMap = toMap(t);
  let counter = t.length;
  let l = 0;
  let i = 0;
  const queue = [];
  while (i < s.length && !tMap[s[i]]) i++;
  l = i;
  while (i < s.length) {
    if (tMap[s[i]] !== undefined) {
      if (i !== l) queue.push(i);
      if (tMap[s[i]]-- > 0) counter--;
      while (counter === 0) {
        if (answer[1] - answer[0] + 1 > i - l + 1) answer = [l, i];
        if (++tMap[s[l]] > 0) counter++;
        l = queue.shift();
      }
    }
    i++;
  }
  return answer[1] === 1000000 ? "" : s.substring(answer[0], answer[1] + 1);
};

const toMap = (string) => {
  const map = {};
  for (const character of string) {
    if (map[character]) {
      map[character]++;
    } else map[character] = 1;
  }
  return map;
};
